pkg/serviceaccount/externaljwt/plugin/keycache.go

```diff
@@ -152,35 +146,57 @@ func (p *keyCache) findKeyForKeyID(keyID string) ([]serviceaccount.PublicKey, bo
 
 // sync supported external keys.
 // completely re-writes the set of supported keys.
-func (p *keyCache) syncKeys(ctx context.Context) (time.Time, error) {
-	val, err, _ := p.syncGroup.Do("", func() (any, error) {
+func (p *keyCache) syncKeys(ctx context.Context) error {
+	_, err, _ := p.syncGroup.Do("", func() (any, error) {
+		oldPublicKeys := p.verificationKeys.Load()
 		newPublicKeys, err := p.getTokenVerificationKeys(ctx)
 		externaljwtmetrics.RecordFetchKeysAttempt(err)
 		if err != nil {
 			return nil, fmt.Errorf("while fetching token verification keys: %w", err)
 		}
 
 		p.verificationKeys.Store(newPublicKeys)
-
 		externaljwtmetrics.RecordKeyDataTimeStamp(newPublicKeys.DataTimestamp.Unix())
 
+		if keysChanged(oldPublicKeys, newPublicKeys) {
+			p.broadcastUpdate()
+		}
+
 		return newPublicKeys, nil
 	})
-	if err != nil {
-		return time.Time{}, err
-	}
-
-	vk := val.(*VerificationKeys)
+	return err
+}
 
-	return vk.DataTimestamp, nil
+// keysChanged returns true if the data timestamp, key count, order of key ids or excludeFromOIDCDiscovery indicators
+func keysChanged(oldPublicKeys, newPublicKeys *VerificationKeys) bool {
+	// If the timestamp changed, we changed
+	if !oldPublicKeys.DataTimestamp.Equal(newPublicKeys.DataTimestamp) {
+		return true
+	}
+	// Avoid deepequal checks on key content itself.
+	// If the number of keys changed, we changed
+	if len(oldPublicKeys.Keys) != len(newPublicKeys.Keys) {
+		return true
+	}
+	// If the order, key id, or oidc discovery flag changed, we changed.
+	for i := range oldPublicKeys.Keys {
+		if oldPublicKeys.Keys[i].KeyID != newPublicKeys.Keys[i].KeyID {
```

Comment:
shall we short before doing this?

Reply:
no, if the backend changes the key order, I'm fine propagating that as a change